pkg/sources/github/feedlet.go

```diff
@@ -0,0 +1,452 @@
+
+	// First, check if service exists already.
+	if sc, err := sc.Get(serviceName, metav1.GetOptions{}); err != nil {
+		if !apierrs.IsNotFound(err) {
+			return nil, fmt.Errorf("service.Get for %q failed: %v", serviceName, err)
+		}
+		log.Printf("service %q doesn't exist, creating", serviceName)
+	} else {
+		log.Printf("found existing service %q", serviceName)
+		// Don't try again.
+		return sc, nil
+	}
+
+	secretName, err := stringFrom(trigger.Parameters, secretNameKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get secret name from trigger parameters: %v", err)
+	}
+
+	secretKey, err := stringFrom(trigger.Parameters, secretKeyKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get secret key from trigger parameters: %v", err)
+	}
+
+	service := resources.MakeService(t.feedNamespace, serviceName, t.image, t.feedServiceAccountName, target, secretName, secretKey)
+	service, err = sc.Create(service)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create service: %s", err)
+	}
+
+	return service, nil
+}
+
+func (t *githubEventSource) deleteReceiveAdapter(namespace string, serviceName string) error {
+	sc := t.servingclientset.ServingV1alpha1().Services(t.feedNamespace)
+
+	// First, check if deployment exists already.
+	if _, err := sc.Get(serviceName, metav1.GetOptions{}); err != nil {
+		if !apierrs.IsNotFound(err) {
+			return fmt.Errorf("services.Get for %q failed: %v", serviceName, err)
+		}
+		log.Printf("service %q already deleted", serviceName)
+		// Don't try again.
+		return nil
+	}
+
+	return sc.Delete(serviceName, &metav1.DeleteOptions{})
+}
+
+func (t *githubEventSource) deleteWebhook(trigger sources.EventTrigger, feedContext sources.FeedContext) error {
+	serviceName := receiveAdapterName(trigger)
+	err := t.deleteReceiveAdapter(t.feedNamespace, serviceName)
+	if err != nil {
+		log.Printf("Error: Failed to delete the ReceiveAdapter \"%s/%s\": %v", t.feedNamespace, serviceName, err)
+		// Continue deleting the webhook.
+	}
+
+	owner, repo, err := parseOwnerRepoFrom(trigger.Resource)
+	if err != nil {
+		log.Printf("Error: Failed to parse owner and repo from tigger.resource: %v; bailing...", err)
+		// Don't try again.
+		return nil
+	}
+
+	webhookID, err := webhookIDFrom(feedContext)
+	if err != nil {
+		log.Printf("Error: Failed to get webhook id from context: %v; bailing...", err)
+		// Don't try again.
+		return nil
+	}
+
+	accessToken, err := stringFrom(trigger.Parameters, accessTokenKey)
+	if err != nil {
+		log.Printf("Error: Failed to get access token from trigger parameters: %v; bailing...", err)
+		// Don't try again.
+		return nil
+	}
+
+	ctx := context.Background()
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: accessToken},
+	)
+	tc := oauth2.NewClient(ctx, ts)
+	client := ghclient.NewClient(tc)
+
+	_, err = client.Repositories.DeleteHook(ctx, owner, repo, webhookID)
+	if err != nil {
+		if errResp, ok := err.(*ghclient.ErrorResponse); ok && errResp.Message == "Not Found" {
```

Comment:
From an earlier comment:
> @evankanderson
> >It looks like the docs for GitHub suggest this should be "missing": https://developer.github.com/v3/#client-errors
> > More generally, string matching like this makes me nervous. Looking at errResp.Response.Status would feel slightly cleaner to me.
> @n3wscott
> @grantr helped with this. looks like github does not use 404, and errors are an array.

Can you put a comment here so future authors aren't confused.